### PR TITLE
feat: Wearable Security Bodycams

### PR DESCRIPTION
**Pull request by Jaimex-3**
## Pull Request Hakkında

Bu PR oyuna Secler için giyilebilir bodycam ekler.

Bodycamler SecTech vendorundan alınabilir, üniformaya aksesuar olarak takılabilir ve kamera konsolları üzerinden izlenebilir. Bodycam takan kişi isterse kamerasını açıp kapatabilir. Birisi bodycam görüntüsünü izlediğinde de kamerayı taşıyan kişiye bununla ilgili bir uyarı gösterilir.

SecTech vendorunda bodycamler: 

image

Bodycam üniformaya takıldığında görünüşü:

image

Bodycam açma/kapatma tuşu:

image

Kamera konsolundan bodycam izleme ve izlenen kişinin gördüğü uyarı:

image

## Oyun İçin Neden Gerekli

Security’nin istasyon içindeki takibini kolaylaştırır. Warden ve AI için ayrı bi güzellik sağlar.

## Changelog

:cl: Jaimex
add: SecTech’e üniformalara aksesuar olarak takılabilen bodycam eklendi.
qol: Bodycam izlenirken kullanıcıya izlendiğini gösteren bir uyarı eklendi.
image: Bodycam için yeni item ve worn iconları eklendi.
/:cl:

### CODE REVIEW

File: code/game/machinery/camera/bodycam_camera.dm [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	clear_watchers()
+	bodycam_component = null
+	return ..()
+
+/obj/machinery/camera/bodycam/on_start_watching(datum/source)
+	if(can_use())
+		var/mob/living/host = loc
+		if(isliving(host))
+			host.throw_alert(ALERT_BODYCAM_VIEWED, /atom/movable/screen/alert/bodycam_viewed)
+	bodycam_component?.on_watch_start(source)
```
Comment (Seefaaa) [inline code/game/machinery/camera/bodycam_camera.dm:32]:
```suggestion
/obj/machinery/camera/bodycam/on_start_watching(datum/source)
	if(can_use())
		astype(loc, /mob/living)?.throw_alert(ALERT_BODYCAM_VIEWED, /atom/movable/screen/alert/bodycam_viewed)
	bodycam_component?.on_watch_start(source)
```

File: code/game/machinery/camera/bodycam_camera.dm [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+/obj/machinery/camera/bodycam/on_stop_watching(datum/no_longer_watching)
+	bodycam_component?.on_watch_stop(no_longer_watching)
+	var/has_live_watcher = bodycam_component && !QDELETED(bodycam_component) && bodycam_component.has_live_watchers()
+	if(!has_live_watcher)
+		var/mob/living/host = loc
+		if(isliving(host) && host.has_alert(ALERT_BODYCAM_VIEWED))
+			host.clear_alert(ALERT_BODYCAM_VIEWED)
```
Comment (Seefaaa) [inline code/game/machinery/camera/bodycam_camera.dm]:
```suggestion
/obj/machinery/camera/bodycam/on_stop_watching(datum/no_longer_watching)
	if(!bodycam_component?.has_live_watchers())
		clear_alert()
	bodycam_component?.on_watch_stop(no_longer_watching)
```

File: code/game/machinery/camera/bodycam_camera.dm [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+	SScameras.remove_camera_from_chunk(src)
+	if(myarea)
+		LAZYREMOVE(myarea.cameras, src)
+		myarea = null
+
+/obj/machinery/camera/bodycam/Destroy()
+	clear_watchers()
+	bodycam_component = null
+	return ..()
```
Comment (Seefaaa) [inline code/game/machinery/camera/bodycam_camera.dm:27]:
```suggestion
/obj/machinery/camera/bodycam/Destroy()
	clear_alert()
	bodycam_component = null
	return ..()
```

File: code/game/machinery/camera/bodycam_camera.dm
Comment (Seefaaa) [inline code/game/machinery/camera/bodycam_camera.dm:1]: loc'u sürekli cast etmek yerine saklayabilirsin de
bknz. [silicon_camera.dm#L7](https://github.com/psychonaut-station/PsychonautStation/blob/master/code/game/machinery/camera/silicon_camera.dm#L7)

File: code/modules/clothing/under/_under.dm [hunk 4]
```diff
@@ -130,6 +130,9 @@
 
 /obj/item/clothing/under/item_interaction(mob/living/user, obj/item/tool, list/modifiers)
 	if(istype(tool, /obj/item/stack/cable_coil))
+		for(var/obj/item/clothing/accessory/bodycam/bodycam in attached_accessories)
+			if(bodycam.broken && bodycam.repair_with_cable(user, tool))
+				return ITEM_INTERACT_SUCCESS
```
Comment (Seefaaa) [inline code/modules/clothing/under/_under.dm]:
```suggestion
		for(var/obj/item/clothing/accessory/bodycam/bodycam in attached_accessories)
			if(bodycam.broken)
				return bodycam.repair_with_cable(user, tool) ? ITEM_INTERACT_SUCCESS : ITEM_INTERACT_BLOCKING
```
tamir edebilse de edemese de devamını engellemesi gerekli

File: code/modules/clothing/under/accessories/bodycam.dm [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+	sync_component_state()
+
+/obj/item/clothing/accessory/bodycam/proc/do_break()
+	broken = TRUE
+	icon_state = "bodycamera_broken"
+	update_appearance()
+	var/obj/item/clothing/under/uniform = loc
+	if(istype(uniform))
+		uniform.update_accessory_overlay()
+	visible_message(span_warning("[src] sparks and powers down!"))
```
Comment (Seefaaa) [inline code/modules/clothing/under/accessories/bodycam.dm:80]:
```suggestion
/obj/item/clothing/accessory/bodycam/emp_act(severity)
	. = ..()
	if(. & EMP_PROTECT_SELF || broken)
		return
	emp_break()
	update_component()

/obj/item/clothing/accessory/bodycam/proc/on_emp(datum/source, severity, protection)
	SIGNAL_HANDLER
	if(protection & EMP_PROTECT_SELF || broken)
		return
	emp_break()
	update_component()

/obj/item/clothing/accessory/bodycam/proc/emp_break()
	broken = TRUE
	icon_state = "bodycamera_broken"
	update_appearance(UPDATE_ICON_STATE)
	astype(loc, /obj/item/clothing/under/uniform)?.update_accessory_overlay()
	visible_message(span_warning("[src] sparks and powers down!"))
```

File: code/modules/clothing/under/accessories/bodycam.dm [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+/obj/item/clothing/accessory/bodycam/proc/sync_component_state()
+	var/obj/item/clothing/under/uniform = loc
+	var/mob/living/wearer = istype(uniform) ? uniform.loc : null
+	if(!istype(wearer))
+		return
+	var/datum/component/pausable_bodycam/component = wearer.GetComponent(/datum/component/pausable_bodycam)
+	if(component)
+		component.set_broken(broken, camera_on)
```
Comment (Seefaaa) [inline code/modules/clothing/under/accessories/bodycam.dm]:
```suggestion
/obj/item/clothing/accessory/bodycam/proc/update_component()
	var/obj/item/clothing/under/uniform = astype(loc)
	var/mob/living/wearer = uniform?.loc
	if(!istype(wearer))
		return
	var/datum/component/pausable_bodycam/component = wearer.GetComponent(/datum/component/pausable_bodycam)
	if(component)
		component.set_broken(broken, camera_on)
```

File: code/modules/clothing/under/accessories/bodycam.dm [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+		if(user)
+			balloon_alert(user, "need cable!")
+		return FALSE
+	broken = FALSE
+	icon_state = "bodycamera"
+	update_appearance()
+	var/obj/item/clothing/under/uniform = loc
+	if(istype(uniform))
+		uniform.update_accessory_overlay()
+	sync_component_state()
```
Comment (Seefaaa) [inline code/modules/clothing/under/accessories/bodycam.dm]:
```suggestion
	update_appearance(UPDATE_ICON_STATE)
	astype(loc, /obj/item/clothing/under/uniform)?.update_accessory_overlay()
	update_component()
```

File: code/modules/clothing/under/accessories/bodycam.dm
Comment (Seefaaa) [inline code/modules/clothing/under/accessories/bodycam.dm:1]: aksesuar takılıyken hem emp_act hem de signal ikisi birden çağırılmıyor değil mi?

File: code/modules/clothing/under/accessories/bodycam.dm [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+	if(user)
+		balloon_alert(user, "repaired")
+	return TRUE
+
+/obj/item/clothing/accessory/bodycam/attackby(obj/item/item, mob/user, list/modifiers)
+	if(broken && istype(item, /obj/item/stack/cable_coil))
+		var/obj/item/stack/cable_coil/cabling = item
+		if(repair_with_cable(user, cabling))
+			return TRUE
+	return ..()
```
Comment (Seefaaa) [inline code/modules/clothing/under/accessories/bodycam.dm:113]:
```suggestion
/obj/item/clothing/accessory/bodycam/attackby(obj/item/item, mob/user, list/modifiers)
	if(broken && istype(item, /obj/item/stack/cable_coil))
		repair_with_cable(user, item)
		return TRUE
	return ..()
```

File: code/datums/components/pausable_bodycam.dm [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+	RegisterSignals(bodycam, list(COMSIG_QDELETING, COMSIG_MOVABLE_MOVED), PROC_REF(camera_gone))
+
+/datum/component/pausable_bodycam/Destroy()
+	force_pause()
+	if(bodycam && !QDELETED(bodycam))
+		UnregisterSignal(bodycam, list(COMSIG_QDELETING, COMSIG_MOVABLE_MOVED))
+		bodycam.bodycam_component = null
+		QDEL_NULL(bodycam)
+	return ..()
```
Comment (Seefaaa) [inline code/datums/components/pausable_bodycam.dm:38]:
```suggestion
/datum/component/pausable_bodycam/Destroy()
	force_pause()
	if(!QDELETED(bodycam))
		UnregisterSignal(bodycam, list(COMSIG_QDELETING, COMSIG_MOVABLE_MOVED))
		bodycam.bodycam_component = null
		QDEL_NULL(bodycam)
	return ..()
```

File: code/datums/components/pausable_bodycam.dm [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+	RegisterSignal(parent, COMSIG_MOVABLE_MOVED, PROC_REF(update_cam))
+	do_update_cam()
+
+/datum/component/pausable_bodycam/proc/on_watch_stop(datum/source)
+	prune_sources_watching(source)
+	if(has_live_watchers())
+		return
+	UnregisterSignal(parent, COMSIG_MOVABLE_MOVED)
+	if(bodycam && !QDELETED(bodycam))
+		SScameras.remove_camera_from_chunk(bodycam)
```
Comment (Seefaaa) [inline code/datums/components/pausable_bodycam.dm]:
```suggestion
	if(!QDELETED(bodycam))
		SScameras.remove_camera_from_chunk(bodycam)
```

File: code/datums/components/pausable_bodycam.dm [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+/datum/component/pausable_bodycam/proc/has_live_watchers()
+	for(var/datum/weakref/weak_source as anything in sources_watching)
+		if(weak_source?.resolve())
+			return TRUE
+	return FALSE
+
+/datum/component/pausable_bodycam/proc/update_cam(datum/source, atom/old_loc, ...)
+	SIGNAL_HANDLER
+	if(get_turf(old_loc) != get_turf(parent))
+		do_update_cam(old_loc)
```
Comment (Seefaaa) [inline code/datums/components/pausable_bodycam.dm]:
```suggestion
/datum/component/pausable_bodycam/proc/on_moved(datum/source, atom/old_loc, ...)
	SIGNAL_HANDLER
	if(get_turf(old_loc) != get_turf(parent))
		update_camera(old_loc)
```

File: code/datums/components/pausable_bodycam.dm [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+/datum/component/pausable_bodycam/proc/on_watch_start(datum/source)
+	prune_sources_watching()
+	var/had_live_watchers = has_live_watchers()
+	LAZYADD(sources_watching, WEAKREF(source))
+	if(had_live_watchers)
+		return
+	if(!bodycam?.camera_enabled)
+		return
+	RegisterSignal(parent, COMSIG_MOVABLE_MOVED, PROC_REF(update_cam))
+	do_update_cam()
```
Comment (Seefaaa) [inline code/datums/components/pausable_bodycam.dm]:
```suggestion
	RegisterSignal(parent, COMSIG_MOVABLE_MOVED, PROC_REF(on_moved))
	update_camera()
```

File: code/datums/components/pausable_bodycam.dm [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+	if(get_turf(old_loc) != get_turf(parent))
+		do_update_cam(old_loc)
+
+/datum/component/pausable_bodycam/proc/do_update_cam(atom/old_loc)
+	if(!bodycam || QDELETED(bodycam))
+		return
+	if(!bodycam.can_use())
+		return
+	SScameras.camera_moved(bodycam, get_turf(old_loc), get_turf(bodycam), camera_update_time)
+	notify_watchers_refresh()
```
Comment (Seefaaa) [inline code/datums/components/pausable_bodycam.dm]:
```suggestion
/datum/component/pausable_bodycam/proc/update_camera(atom/old_loc)
	if(QDELETED(bodycam))
		return
	if(!bodycam.can_use())
		return
	SScameras.camera_moved(bodycam, get_turf(old_loc), get_turf(bodycam), camera_update_time)
	notify_watchers_refresh()
```

File: code/datums/components/pausable_bodycam.dm [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+		return
+	bodycam.camera_enabled = !broken && camera_on
+	if(broken || !bodycam.camera_enabled)
+		force_pause()
+
+/datum/component/pausable_bodycam/proc/notify_watchers_disconnect()
+	if(!bodycam || QDELETED(bodycam))
+		return
+	var/list/to_notify = length(sources_watching) ? sources_watching.Copy() : list()
+	for(var/datum/weakref/weak_source as anything in to_notify)
```
Comment (Seefaaa) [inline code/datums/components/pausable_bodycam.dm]:
```suggestion
/datum/component/pausable_bodycam/proc/notify_watchers_disconnect()
	if(QDELETED(bodycam))
		return
	for(var/datum/weakref/weak_source as anything in sources_watching)
```
iterasyonda listeyi değiştirmeyeceksen kopyalamana gerek yok

File: code/datums/components/pausable_bodycam.dm [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+			var/datum/computer_file/program/secureye/program = source
+			program.update_active_camera_screen()
+
+/datum/component/pausable_bodycam/proc/force_pause()
+	notify_watchers_disconnect()
+	LAZYCLEARLIST(sources_watching)
+	UnregisterSignal(parent, COMSIG_MOVABLE_MOVED)
+	if(bodycam && !QDELETED(bodycam))
+		bodycam.clear_watchers()
+		SScameras.remove_camera_from_chunk(bodycam)
```
Comment (Seefaaa) [inline code/datums/components/pausable_bodycam.dm:149]:
```suggestion
/datum/component/pausable_bodycam/proc/force_pause()
	notify_watchers_disconnect()
	LAZYNULL(sources_watching)
	UnregisterSignal(parent, COMSIG_MOVABLE_MOVED)
	if(!QDELETED(bodycam))
		bodycam.clear_alert()
		SScameras.remove_camera_from_chunk(bodycam)
```

File: code/game/machinery/camera/bodycam_camera.dm [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+		var/mob/living/host = loc
+		if(isliving(host) && host.has_alert(ALERT_BODYCAM_VIEWED))
+			host.clear_alert(ALERT_BODYCAM_VIEWED)
+
+/obj/machinery/camera/bodycam/proc/clear_watchers()
+	var/mob/living/host = loc
+	if(!isliving(host))
+		return
+	if(host.has_alert(ALERT_BODYCAM_VIEWED))
+		host.clear_alert(ALERT_BODYCAM_VIEWED)
```
Comment (Seefaaa) [inline code/game/machinery/camera/bodycam_camera.dm]:
```suggestion
/obj/machinery/camera/bodycam/proc/clear_alert()
	astype(loc, /mob/living)?.clear_alert(ALERT_BODYCAM_VIEWED)
```

File: code/game/machinery/camera/bodycam_camera.dm
Comment (Jaimex-3) [reply code/game/machinery/camera/bodycam_camera.dm:1]: hallettim

File: code/modules/clothing/under/accessories/bodycam.dm
Comment (Jaimex-3) [reply code/modules/clothing/under/accessories/bodycam.dm:1]: Hayır emp uniforma geliyor, uniformdan bodycame signal gidiyor.

**Comment by Jaimex-3** [comment]
eline sağlık biraz uğraştırdım kusura bakma.

File: code/game/machinery/computer/camera.dm [hunk 17]
```diff
@@ -19,6 +19,11 @@
 	// Stuff needed to render the map
 	var/atom/movable/screen/map_view/camera/cam_screen
 
+/obj/machinery/computer/security/proc/release_active_camera()
+	active_camera?.on_stop_watching(src)
+	active_camera = null
+	last_camera_turf = null
+
```
Comment (Seefaaa) [inline code/game/machinery/computer/camera.dm]: dosyada uygun bir yere koy `/atom/movable/screen/map_view/camera` (173. satır) öncesine mesela

File: code/modules/unit_tests/bodycam_watchers.dm [hunk 18]
```diff
@@ -81,6 +82,77 @@
...
+	camera.on_start_watching(console)
+
+	TEST_ASSERT(host.has_alert(ALERT_BODYCAM_VIEWED), "Host should gain the viewed alert while a security console is watching.")
+
+	console.ui_close(viewer)
+
+	TEST_ASSERT(!component.has_live_watchers(), "Closing the security console should remove the live watcher.")
+	TEST_ASSERT(!host.has_alert(ALERT_BODYCAM_VIEWED), "Closing the security console should clear the watched alert.")
+
+/datum/unit_test/bodycam_watchers_console_close_dedupes_users/Run()
```
Comment (Seefaaa) [inline code/modules/unit_tests/bodycam_watchers.dm]: neyi test ediyor bu test

Comment (Jaimex-3) [reply code/modules/unit_tests/bodycam_watchers.dm]: alert hata olup 2 kere oluşursa izlenme bırakıldığında silinmesi için.

Comment (Seefaaa) [reply code/modules/unit_tests/bodycam_watchers.dm]: bi üstündeki test ile tek farkı
```diff
- console.concurrent_users += REF(viewer)
+ console.concurrent_users |= viewer_ref
+ console.concurrent_users |= viewer_ref
+ TEST_ASSERT_EQUAL(length(console.concurrent_users), 1, "Security console should only track a viewer once.")
```
değil mi nasıl gerçekleşecek o?
`|=` operatörünü mü sınıyorsın?

Comment (Jaimex-3) [reply code/modules/unit_tests/bodycam_watchers.dm]: evet üsttekinde 1 kere viewer ekliyo alttakinde 2 kere ekliyor

Comment (Seefaaa) [reply code/modules/unit_tests/bodycam_watchers.dm]: `|=` 2 kere nasıl eklesin?

Comment (Seefaaa) [reply code/modules/unit_tests/bodycam_watchers.dm]: yaptığın şey `assert_eq(1 + 1, 2)` gibi bir şey

Comment (Jaimex-3) [reply code/modules/unit_tests/bodycam_watchers.dm]: evet anlamlı olması için += olması lazım dı şuan anlamı yok testin

Comment (Seefaaa) [reply code/modules/unit_tests/bodycam_watchers.dm]: `+=` ile değiştirsen de anlamsız bir test olurdu yine

Comment (Jaimex-3) [reply code/modules/unit_tests/bodycam_watchers.dm]: fail döncek dur bakayım ona

Comment (Jaimex-3) [reply code/modules/unit_tests/bodycam_watchers.dm]: o test anlamsız kalıyor direkt kaldırıyorum

**Comment by Seefaaa** [comment]
izleyen ghost atınca veya silinince bodycam alerti kalmaya devam ediyo

bu edge caseleri kendin denesene localde

File: code/game/machinery/computer/camera.dm [hunk 19]
```diff
@@ -57,8 +58,8 @@
 		var/is_living = isliving(user)
 		// Ghosts shouldn't count towards concurrent users, which produces
 		// an audible terminal_on click.
-		if(is_living)
-			concurrent_users += user_ref
+		if(is_living && !(user_ref in concurrent_users))
+			concurrent_users |= user_ref
```
Comment (Seefaaa) [inline code/game/machinery/computer/camera.dm]:
```suggestion
		if(is_living)
			concurrent_users |= user_ref
```
`|=`'in olayı bu zaten

File: code/game/machinery/computer/camera.dm [hunk 20]
```diff
@@ -153,12 +164,34 @@
...
+	active_camera?.on_stop_watching(src)
+	active_camera = null
+	last_camera_turf = null
+
+/obj/machinery/computer/security/proc/has_living_viewers(mob/excluding_user)
+	if(!LAZYLEN(open_uis))
+		return FALSE
+	for(var/datum/tgui/ui as anything in open_uis)
+		if(!istype(ui, /datum/tgui))
+			continue
```
Comment (Seefaaa) [inline code/game/machinery/computer/camera.dm]:
```suggestion
	for(var/datum/tgui/ui in open_uis)
```
typecheck yapacaksan `as anything` kullanma zaten

File: code/modules/modular_computers/file_system/programs/secureye.dm [hunk 21]
```diff
@@ -101,8 +105,8 @@
 	var/is_living = isliving(user)
 	// Ghosts shouldn't count towards concurrent users, which produces
 	// an audible terminal_on click.
-	if(is_living)
-		concurrent_users += user_ref
+	if(is_living && !(user_ref in concurrent_users))
+		concurrent_users |= user_ref
```
Comment (Seefaaa) [inline code/modules/modular_computers/file_system/programs/secureye.dm]:
```suggestion
	if(is_living)
		concurrent_users |= user_ref
```
yukarıda söyledim

File: code/modules/modular_computers/file_system/programs/secureye.dm [hunk 22]
```diff
@@ -195,17 +199,60 @@
...
 
+/datum/computer_file/program/secureye/proc/has_living_viewers(mob/excluding_user)
+	var/list/viewer_uis = computer?.open_uis
+	if(!LAZYLEN(viewer_uis))
+		viewer_uis = open_uis
+	if(!LAZYLEN(viewer_uis))
+		return FALSE
+	for(var/datum/tgui/ui as anything in viewer_uis)
+		if(!istype(ui, /datum/tgui))
+			continue
```
Comment (Seefaaa) [inline code/modules/modular_computers/file_system/programs/secureye.dm]:
```suggestion
	for(var/datum/tgui/ui in viewer_uis)
```
yukarıda söylediö

**Comment by Seefaaa** [comment]
testlerin sorununu düzeltmeni bekliyorum

image
image

**Comment by Jaimex-3** [comment]
> testlerin sorununu düzeltmeni bekliyorum
> 
> image
> image
> 

Finallerden dolayı yavaşladım haftasonu halledicem.

**Review by Seefaaa — APPROVED** [review]
eline sağlık

**Comment by Jaimex-3** [comment]
> eline sağlık

senin de eline sağlık.
